Step5 자동차 경주(리팩터링)

jaeykweon opened the pull request:
5단계 자동차 경주(리팩터링) 리뷰 요청드립니다!

4단계 마지막에 주신 리뷰 내용 반영하였습니다

다만 5단계 요구사항을 적용함에 있어서

단순히 main함수의 내용을 controller로 옮긴 것 밖에는 없는데

또 따로 적용해야 할 부분이 있는지 질문드립니다

jaeykweon committed:
refactor: [4단계] 코드리뷰 적용

- Winners 클래스 생성
- nextRound 메소드의 반환값 변경
- printResult를 각 일급 컬렉션으로 넘김

jaeykweon committed:
refactor: [4단계] 패키지 이름 변경

jaeykweon committed:
feature: [5단계] controller 구현

jaeykweon committed:
config: [5단계] upgrade kotest runner version

ohgillwhan commented on src/main/kotlin/carrace/Controller.kt:
```diff
@@ -0,0 +1,22 @@
+
+    fun startGame() {
+        val cars = InputView.inputCars()
+        val gameRound = InputView.inputGameRound()
```
`InputView`, `OutputView`도 `Interface`를 통해 추상화를 한다면, `Controller`도 테스트가 가능해질거에요.

ohgillwhan commented on src/main/kotlin/carrace/logic/car/Cars.kt:
```diff
@@ -0,0 +1,36 @@
+            cars.filter { car ->
+                car.carPosition.position == highestScore
+            }
+        )
```
이 로직은 `Winner`로 옮겨서 팩터리 메서드 등을 통해서 구현하는 건 어떨까요?

ohgillwhan commented on src/main/kotlin/carrace/logic/car/Car.kt:
```diff
@@ -0,0 +1,20 @@
+) {
+    fun nextRound(movingSystem: MovingSystem): Car {
+        val movingDistance = movingDistance(movingSystem)
+        return Car(
+            this.carName,
+            CarPosition(this.carPosition.position + movingDistance)
+        )
```
자동차는 게임상에서 이름이라는 식별자를 가지고 있는, `Entity`라고 봐도 괜찮지 않을까요? 새롭게 만드는것보다는 자동차 정보(CarInfo Class등을 만들어서)를 반환하는건 어떨까요?
엔티티는 자기의 생명주기가 있고, 그것을 가지고 가야합니다!

ohgillwhan commented on src/main/kotlin/carrace/logic/car/Car.kt:
```diff
@@ -0,0 +1,20 @@
+        val movingDistance = movingDistance(movingSystem)
+        return Car(
+            this.carName,
+            CarPosition(this.carPosition.position + movingDistance)
```
`carPosition`에 함수를 만들어 놓으셨네요~

```kotlin
carPosition.move(movingDistance)
```

ohgillwhan commented on src/main/kotlin/carrace/logic/car/Cars.kt:
```diff
@@ -0,0 +1,36 @@
+        )
+    }
+
+    fun getResult(): List<String> {
+        return cars.map { car ->
+            "${car.carName.name} : " + "-".repeat(car.carPosition.position + 1)
+        }
+    }
```
화면에 대한 로직이네요~ `ResultView`가 더 잘 어울릴거 같아요!

ohgillwhan commented on src/main/kotlin/carrace/logic/car/Winners.kt:
```diff
@@ -0,0 +1,11 @@
+data class Winners(
+    private val cars: List<Car>
+) {
+    fun getResult(): String {
+        return "${cars.joinToString { car ->
+            car.carName.name
+        }}가 최종 우승했습니다."
+    }
```
화면에 대한 로직이네요~ `ResultView`가 더 잘 어울릴거 같아요!

ohgillwhan commented on src/main/kotlin/carrace/logic/car/Cars.kt:
```diff
@@ -0,0 +1,36 @@
+    )
+
+    fun nextRound(movingSystem: MovingSystem): Cars {
+        return Cars(
+            cars.map { car ->
+                car.nextRound(movingSystem)
+            }
+        )
```
만약 `Car`에서 `CarInfo`를 반환한다면. 일급컬렉션을 계속 만들지 않아도 될것 같아요.

ohgillwhan reviewed (changes requested):
재용님 몇가지 고민해볼 리뷰 남겨봤습니다!
확인 부탁드려요~

jaeykweon committed:
refactor: [5단계] 코드리뷰 적용

- carInfoDto 생성
  - 도메인 영역 밖에서는 프로퍼티 직접 접근 못하게 변경
- view에 관련한 부분은 전부 view로 위임
- view에 인터페이스 적용해서 추상화
- 기존 테스트 코드 수정
- 컨트롤러 테스트 추가

jaeykweon commented:
@ohgillwhan 말씀하신 부분들 [refactor: [5단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/931/commits/84e4bc831c6e23648ec979006acff6cba0568485)에 적용해 보았습니다
피드백 부탁드립니다!!

ohgillwhan commented on src/main/kotlin/carrace/Controller.kt:
```diff
@@ -0,0 +1,32 @@
+
+        val gameLog = mutableListOf(CarInfos(cars))
+
+        if (gameRound.value > 1) {
```
`gameRound`도 객체이니, `.value > 1` 보다는 더 의미있는 이름의 메서드를 뽑을 수 있겠네요~

ohgillwhan commented on src/main/kotlin/carrace/logic/car/Cars.kt:
```diff
@@ -0,0 +1,25 @@
+        }
+    )
+
+    fun getWinners(): Winners {
+        return Winners.from(cars)
+    }
```
👍

ohgillwhan commented on src/main/kotlin/carrace/view/ResultViewImpl.kt:
```diff
@@ -0,0 +1,24 @@
+    override fun displayGameResult(gameLog: List<CarInfos>) {
+        println("실행 결과")
+        gameLog.forEach { carInfos ->
+            carInfos.carInfos.forEach { carInfo ->
+                println("${carInfo.carName.name} : " + "-".repeat(carInfo.carPosition.position + 1))
+            }
+            println()
```
`indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.`라는 조건을 한번 만족해볼까요?

ohgillwhan commented on src/main/kotlin/carrace/Controller.kt:
```diff
@@ -0,0 +1,32 @@
+        val cars = inputView.inputCars()
+        val gameRound = inputView.inputGameRound()
+
+        val gameLog = mutableListOf(CarInfos(cars))
+
+        if (gameRound.value > 1) {
+            for (i in 1..gameRound.value) {
+                gameLog.add(
+                    cars.nextRound(RandomMovingSystem)
+                )
+            }
+        }
```
코틀린에서 최대한 불변할 수 있게 (mutable 최대한 사용 안하기!) 작업을 해보시면 더 재밌을거에요!
```kotlin
        val gameLog: List<CarInfos> = listOf(CarInfos(cars)) + if (gameRound.value > 1) {
            (1..gameRound.value).map { cars.nextRound(RandomMovingSystem) }
        } else {
            emptyList()
        }
```

ohgillwhan reviewed (approved):
재용님 코드 잘 봤습니다.
사소한 리뷰 남겨봤는데, 한번 고민해보시고 궁금한 점은 DM 주세요!
고생하셨습니다!!
